v2/robotmk/scheduler.py

```diff
@@ -1,96 +1,151 @@
-from typing import Final
+from collections.abc import Iterable, Iterator, Mapping, Sequence
+from typing import Final, assert_never
 from uuid import uuid4
 
 from apscheduler.schedulers.blocking import BlockingScheduler  # type: ignore[import]
 from apscheduler.triggers.interval import IntervalTrigger  # type: ignore[import]
-from pydantic import BaseModel
+from pydantic import BaseModel, TypeAdapter
 from robot import rebot  # type: ignore[import]
 
 from robotmk.environment import RCCEnvironment, ResultCode, RobotEnvironment
 from robotmk.runner import Attempt, RetrySpec, RetryStrategy, Variant, create_attempts
 from robotmk.session import CurrentSession, UserSession
 
 
-class _RCC(BaseModel, frozen=True):
-    binary: pathlib.Path
-    robot_yaml: pathlib.Path
+class _RCCConfig(BaseModel, frozen=True):
+    binary_path: pathlib.Path
 
 
-class _UserSession(BaseModel, frozen=True):
+class _UserSessionConfig(BaseModel, frozen=True):
     user_name: str
 
 
-class _SuiteConfig(BaseModel, frozen=True):  # pylint: disable=too-few-public-methods
+class _SuiteConfig(BaseModel, frozen=True):
     execution_interval_seconds: int
     robot_target: pathlib.Path
-    working_directory: pathlib.Path
     variants: Sequence[Variant]
     retry_strategy: RetryStrategy
-    env: _RCC | None
-    session: _UserSession | None
+    session: _UserSessionConfig | None
+
+
+class _SystemPythonSuiteConfig(_SuiteConfig, frozen=True):
+    ...
+
 
+class _RCCSuiteConfig(_SuiteConfig, frozen=True):
+    robot_yaml_path: pathlib.Path
 
-class _SchedulerConfig(BaseModel, frozen=True):
-    suites: Mapping[str, _SuiteConfig]
 
+class _ConfigSystemPython(BaseModel, frozen=True):
+    rcc: None
+    working_directory: pathlib.Path
+    suites: Mapping[str, _SystemPythonSuiteConfig]
+
+
+class _ConfigRCC(BaseModel, frozen=True):
+    rcc: _RCCConfig
+    working_directory: pathlib.Path
+    suites: Mapping[str, _RCCSuiteConfig]
 
-def _scheduler(suites: Mapping[str, _SuiteConfig]) -> BlockingScheduler:
+
+def _scheduler(config: _ConfigSystemPython | _ConfigRCC) -> BlockingScheduler:
     scheduler = BlockingScheduler()
-    for suite_name, suite_config in suites.items():
+    for (suite_name, suite_config), rcc_env_spec in zip(
+        config.suites.items(),
+        _rcc_env_specs(config),
+    ):
         scheduler.add_job(
-            _SuiteRetryRunner(suite_name, suite_config),
+            _SuiteRetryRunner(
+                suite_name=suite_name,
+                suite_config=suite_config,
+                environment=rcc_env_spec,
+                working_directory=config.working_directory,
+            ),
             name=suite_name,
             trigger=IntervalTrigger(seconds=suite_config.execution_interval_seconds),
             next_run_time=datetime.datetime.now(),
         )
     return scheduler
 
 
+@dataclasses.dataclass(frozen=True)
+class _RCCEnvironmentSpec:
+    binary_path: pathlib.Path
+    robot_yaml_path: pathlib.Path
+
+
+def _rcc_env_specs(
```

Comment:
Probably more appropriate as class method

Reply:
I reworked this, I think it's formulated in a better way now. Good suggestion, thanks!